### ISSUE TITLE
Transposing XML before loading it into OSMD

### ISSUE THREAD

**Issue by AlbertHart** [opensheetmusicdisplay/opensheetmusicdisplay#613]
i wrote a Javascript function to transpose the XML before it is loaded into OSMD.

Basically, it reads the XML - line by line - and then rewrites it, while transposing just 4 entries.

`<fifths>-4</fifths>`

```
<pitch>
     <step>D</step>
     <octave>4</octave>
    </pitch>
```

```
<root>
      <root-step>B</root-step>
     <root-alter>-1</root-alter>
</root>
```

```
<bass>
     <bass-step>F</bass-step>
     <bass-alter>1</bass-alter>
    </bass>
```

There is a little work involved to get #'s vs b's to match key and chords, and to change the octave properly. And I still need to run it through a little more testing.

But I anyone would like a copy of it, let me know. I can post it here if you like.

Or if OSMD wants to look an including it in the full version.

This is going to be great for my project because I can easily transpose and print my scores without having to return to Sibelius, transopose, resave as MusicXML and reload.

#580

**Comment by sschmidTU** [comment]
If you can share your code, that would be great. Do you have a fork of the repository?
Also, if you want to transpose pieces with explicitly given key signatures (at the beginning of the staffline), you'll need to transpose the key instruction too.

**Comment by ice6** [comment]
If I want to implement this feature, where should I start the code?

change the musicxml part or change the omsd object ?

**Comment by AlbertHart** [comment]
The routine I wrote takes the loaded, but not processed, ascii xml text and reads things like <pitch>, and <fifth> and transposes them. If you have a musicxml file you want to send me, I'll dust the off and transpose it. It also attempts to reverse stems, and accidentals, etc.

One use would be as a function people can call after loading the XML ascii file to tarnspose the before processing it. We could add this to the OSMD demo. WHich would then load a musicxml file, but display it transposed. This would be the function we would like to expose to other developers.

Another use would be a stand-alone musicxml transposer. Which would be useful for people using musicxml for other purposes. It is also valuable as a debugging tool for us. For example, I can output musicxml from Sibelius, run it through my routine, and compare it to a second musicxml file which was transposed in sibelius to look for items I may have missed.

If you want to write a similar routine, I'll be glad to send my routine as a starting point. It is in Javascript and not in typescript - which I haven't mastered (or done anything useful with yet)

![image](https://user-images.githubusercontent.com/8965352/81614256-028a7b80-939d-11ea-8ece-2d9af4be8f7d.png)

**Comment by ice6** [comment]
@AlbertHart Hi thanks for your reply, I read the musescore's implementation [here](https://github.com/musescore/MuseScore/blob/master/libmscore/transpose.cpp), it is quite complicated.

It is helpfull if you can send your me the routine as a starting point. :)

And I wonder whether the osmd team is working on this? @sschmidTU

**Comment by AlbertHart** [comment]
Mine is equally complicated - using rules of fifths logic to determine what note to transpose to, etc.

I created a stand-alone test version which you can run at:

http://alhartapps.com/transpose/transpose.htm 

- Set a key
- load a musicxml file

and the transposed text will appear at the bottom.

You can test it by saving the transposed text, and loading it into musescore, etc.
(I will add a save button if we make changes)

You can grab the source and the source of the included .js file from your browser

If you prefer I will upload it here.

Let me know about any problems you see, or things you would like documented and I will take a look at them.

![image](https://user-images.githubusercontent.com/8965352/81732297-8a2fc300-944d-11ea-9ac3-4edb582a4adc.png)

**Comment by AlbertHart** [comment]
The test routine does not need or use and OSMD code - because that is not needed for the trasposition.

**Comment by ice6** [comment]
@AlbertHart 👍 thank you so much, I will test it and come back to share my thoughts.

**Comment by ice6** [comment]
@AlbertHart I tried your routine, and basically it works!!! Wonderful 👍 
I found a small bug : 

`<bass>
<bass-step>transposed_note.step</bass-step>
</bass>`

when the routine encounter `bass-step`, it will not work.

I will test it will more scores tonight.

**Comment by ice6** [comment]
@AlbertHart 

If it is possible, you can upload the code your github repository named MusicXMLTransposer or something similar.

Then I can contribute code to get it done, maybe from a routine to a library for convenient use.

Really appreciate about your work.

**Comment by ice6** [comment]
for `osmd`, can use this library directly on the `load` function, or change osmd object directly to gain better performance without touching musicxml file.

**Comment by AlbertHart** [comment]
**transposed_note.step**

I've been using code like

```
bass_xml = `<bass>
                <bass-step>` + transposed_note.step + `</bass-step>\n`
```;

I'll switch it to 

```
bass_xml = `<bass>
                <bass-step>$transposed_note.step</bass-step>\n`;
```

I was previously using sprintf() to do these, but I didn't want to require a sprintf library include and messed up the changes I made to it.

**Comment by AlbertHart** [comment]
> for `osmd`, can use this library directly on the `load` function, or change osmd object directly to gain better performance without touching musicxml file.

In my own code, I load a musicxml file from the the disk, transpose it, and then call osmd.load()

basically as below.

So it works well with routines which pass a string of ascii xml to osmd.load()

```
xml_string = xml_string_passed;

if (parameters.transpose_key != "None");
        {
            transpose_xml();
        }

osmd = new opensheetmusicdisplay.OpenSheetMusicDisplay(container, {
                backend: backend_type,
                drawingParameters: "compact",
                drawPartNames: false
            });

osmd.load(xml_string);

```

**Comment by AlbertHart** [comment]
> @AlbertHart
> 
> Cound you upload the code to your github repository named MusicXMLTransposer or something similar.
> 
> Then I can contribute code to get it done, maybe from a routine to a library for convenient use.
> 
> Really appreciate about your work.

I don't really have a github repository of anything. I have a github account, but haven't used it for source control of development.

How about the opposite,  if I upload the code here, and you stick it in github somewhere, then we can both work on it there?

Or you can pass some instructions on what I need to do to use github better.

**Comment by AlbertHart** [comment]
I created a new version at:


with the <bass-step> stuff fixed (maybe - it is not used in any of my scores)

I also, added a download button which will save your transposed score to your local drive.
(I need to fix it so you can specify a file name)

I tried pasting the downloaded score to the OSMD demo and it seems to work.

I am now going to remove unused items in the transpose.js file, and also make the transposition stuff an object to avoid name conflicts.

Let me know if you get a chance to test it more.

I'll let you know when I have another new version available.

![image](https://user-images.githubusercontent.com/8965352/81815705-8a7c9c80-94e7-11ea-9c6f-05b5c67c18ce.png)

**Comment by AlbertHart** [comment]
I got the namespace stuff working, so that everything in transpose_functions.js is in an objected called osmd_transpose. And have uploaded it onto my site.

I think we should work together to create a new version of 

https://opensheetmusicdisplay.github.io/demo/

which has options to transpose the song.

Step 1: First, I think we should get a version of /demo/ working using osmd_transpose as a seperate object, and including transpose_functions.js. This also lets me test stuff with the existing transpose.htm - all in ascii - and also lets me download the transposed .musicxml file for easy compoarison to the original file, or files transposed from out sources.

Step2: Then we can take the stuff in transpose_functions.js and put it directly into OSMD.

Since I have no idea how to do this, I propose that you create a version of /demo/ that e can work on together, and then we can insert stuff into it.

We could add something like this, which would transpose the pasted XML and redraw the scod.

![image](https://user-images.githubusercontent.com/8965352/81834681-814afa00-94fe-11ea-97e5-406f6f237220.png)

**Comment by ice6** [comment]
@AlbertHart 👍 wow, cool. I will test is with more score manually. maybe later we can test it automatically.

Integrated into `osmd` is not hard for me.

But at first, we have to make library easier to use, for example:

1. make the code reenterable

2. make the code follow some rules like `[Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle)`

it is not hard :)

After finishing this, the `Step 1` will be finished.

and `Step 2` is easy to implement, modify the raw musicxml and reload it with `osmd` will make it work, but I think it will be not that smooth, but is will work. I can try.

So,  can I upload the your new code to a github repository, so we can track the code.

May we can use [jitsi](https://meet.jit.si/)  to have a video conference for collaborating using github. :)

**Comment by AlbertHart** [comment]
Yes

Can you just grab the code from the url, or do you want me to upload it..

I made another change today - to pass the parameters and the xml string to the transpose_xml function. I’ll check later to make sure I uploaded the latest. 

Al

**Comment by ice6** [comment]
@AlbertHart cool :) that is about [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle)

**Comment by AlbertHart** [comment]
> A small bug: if I want to transpose score to A#, it will not work, because in the accidentals_in_key dictionary, there is no A#. Bb will work.

My Piano teacher advised me to leave A# out, because he said no one would ever use it. Although he agreed that I needed to be able to play scales and chords in the Cb scale..

But it is relatively easy to add.

**Comment by AlbertHart** [comment]
> So, can I upload the your new code to a github repository, so we can track the code.


I just uploaded the latest version which passes the parameters and  xml string into transpose_xml() so that it is not using global variables

**Comment by ice6** [comment]
@AlbertHart maybe we have same teacher, He also told me so. :P

**Comment by AlbertHart** [comment]
If he has the perfect way to handle every question - never giving me a direct answer but rather guiding me to it - then we might.  

And if A#, why not E# - since we do process Cb for other reasons.  What we start to run into with strange keys is whether we are also going to transpose individual notes into double sharps and flats. 

Al

**Comment by ice6** [comment]
this is the screenshot taken from musescore.

![image](https://user-images.githubusercontent.com/889040/81901717-8f2d7780-95f1-11ea-9611-63d718db89ae.png)

```
Cb/Abm
Gb/Ebm
Db/Bbm
Ab/Fm
Eb/Cm
Bb/Gm
F/Dm
C/Am
G/Em
D/Bm
A/F#m
E/C#m
B/G#m
F#/D#m
C#/A#m
```

there are 15 pairs in musescore's option list.

Maybe we can support it as well.

**Comment by matt-uib** [comment]
Dear @ice6 and @AlbertHart as we see now your active discussion here about the transpose feature you drew our attention back on this feature. 
Sorry from our side, that we couldn't make any progress on the transpose feature, but as so often: our resources are quite limited and tasks to do seem unlimited :-)
But as our plan always was to provide a plugin for transposing which we also want to license for commercial solutions using OSMD (it will be free for non-commercial use), we want to have now a working solution for you and others. 
The idea behind this is to provide some income with some premium plugins to get the money needed to allow a sustainable development of OSMD.
As we have already solved the transposing in our native apps (Practice Bird, PhonicScore) we have a working solution that perfectly integrates into OSMD. We have already translated it to typescript and do some final testing and fixing.
The advantage of our solution is, that is integrated into OSMD and works directly within the OSMD object model and therefore is quite fast.
We also transpose chord symbols correctly as well as consider problems like A#.
We will try to have this ready as preview for anyone interested hopefully by tomorrow.

So it would be totally understandable, if you would like to progress now on your own solution, but maybe our solution is also of interest to you or others.

**Comment by ice6** [comment]
@matt-uib 👍 👍 👍 

especially `works directly within the OSMD object model and therefore is quite fast`. 

That is exciting for `We will try to have this ready as preview for anyone interested hopefully by tomorrow.`. Can't wait to see it.

So @AlbertHart , your routine would be better named `MusicXMLTransposer`, because it manipulates `musicxml` file directly, it has more common scenarios and works quite well.

It is thankfull for making `osmd` forward.

**Comment by sschmidTU** [comment]
> My teacher told me there is no need to support Cb/Abm F#/D#m C#/A#m, because no one use it.
But from the engineering perspective, we'd better support it.

I know a few piano pieces in Cb and F#. You should definitely support all the keys.

**Comment by ice6** [comment]
@sschmidTU I never met, not suporting it is ok for me :), I don't want to read double flat or double sharp score :P

But for a software, it is better to support to cover all cases.

**Comment by sschmidTU** [comment]
@ice6 haha, yes, it's uncommon, but it happens.
I've played this piece, Alkan Concerto for solo piano, which is in G# minor, and has a few triple sharps ;)
![image](https://user-images.githubusercontent.com/33069673/81923689-589a3180-95de-11ea-9c0d-ec6ef8da39cf.png)
https://dictionary.onmusic.org/terms/3687-triple_sharp
That's why OSMD supports triple sharps and triple flats, see OSMD Function Test - Accidentals.
![image](https://user-images.githubusercontent.com/33069673/81923761-75366980-95de-11ea-8a0b-26022fc82f2b.png)

**Comment by AlbertHart** [comment]
I'm going to add an option "use double flats" which, if not selected, will replace bb and ## with a different symbol. Cbb --> Bb

**Comment by ice6** [comment]
@AlbertHart cool 👍

**Comment by AlbertHart** [comment]
> I'm also going to add an option "use double flats" which, if not selected, will replace bb and ## with a different symbol. for example: Cbb --> Bb, Ebb --> D

I think I am going to defer this issue until someone requests it later.

Here is a score in MuseScore with lots of alternative note spellings and the transposition by MuseScore with and without using double sharps and flats.

![image](https://user-images.githubusercontent.com/8965352/81959028-1cf66c00-95cc-11ea-9425-5f7e1cab7441.png)

I think our time is better spent getting transposition integrated into OSMD, and not complicating things yet.

**Comment by ice6** [comment]
@AlbertHart I make it work [here](https://github.com/ice6/opensheetmusicdisplay). 

But I don't know how to publish the code because the `osmd`'s structure is complex. I will figure it out.

You can clone the code, and run `npm install` & `npm start` to test the code.

It looks like below:

![image](https://user-images.githubusercontent.com/889040/82078228-5e456380-9713-11ea-844c-f7f719e333fd.png)

![image](https://user-images.githubusercontent.com/889040/82078250-6bfae900-9713-11ea-8546-331a2bda2ada.png)


I test some score, and some target key will fail.

Let us fix it when you are available.

**Comment by sschmidTU** [comment]
it's the main piece in our demo.
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/test/data/Beethoven_AnDieFerneGeliebte.xml

**Comment by AlbertHart** [comment]
Thanks, I'll use it for my own testing as well.

**Comment by ice6** [comment]
After a good sleep, I figure it out how to publish the work as github pages.

[https://ice6.github.io/opensheetmusicdisplay/build/](https://ice6.github.io/opensheetmusicdisplay/build/)

I add a `MusicXML` property to the `osmd` object, so every transpose based on the `transposed` raw `musicxml`. After transposing several times, it will fail.

You can open Chrome Developer Console, use `osmd.musicxml` to get the raw musicxml for debugging purpose.

**Comment by AlbertHart** [comment]
We should keep the original xml_string if possible, so if they transpose again, they start with the original.

We should fix the multiple transposition problem. I can test that easily by transposing  score  twice and comparing to two transpositions of the original score.  I'll take a look at that 明天

**Comment by ice6** [comment]
@AlbertHart I fixed the problem, now it is based upon the original musicxml every transpose.

I add a `window.originalMusicXML` global variable to hold the `original musicxml`, the `osmd.musicxml` is the transposed one.

If you want the successive transpose, set `window.originalMusicXML` undefined will do the trick.

the code:
[https://github.com/ice6/opensheetmusicdisplay](https://github.com/ice6/opensheetmusicdisplay)

the demo:
[https://ice6.github.io/opensheetmusicdisplay/build/](https://ice6.github.io/opensheetmusicdisplay/build/)

**Comment by leo6104** [comment]
@ice6 Thanks for sharing your code.

I refactor your code to typescript & use depth-first traversing DOM method. 

Here is the code i refactored. 
https://gist.github.com/leo6104/85869c82485da716d7274a685d9cb5dc

With this code, i can transpose osmd musicxml like below.
```
const transposer = new MusicXmlTransposeService();
await transposer.load(xmlPath);
transposer.transpose('E');
await this.osmd.load(transposer.xml);
await this.osmd.render();
```

**Comment by AlbertHart** [comment]
leo6104: did you use a routine somewhere to refactor the code?

ice6 - can you get a working copy of the refactored code and put it into out test area, (and make it work)

I can then merge the changes I am making today to the raw Javascript into the new code.

**Comment by ice6** [comment]
@leo6104 @AlbertHart sorry, I just see this discussion.

@AlbertHart is still making the routine works better.

I will have a look at @leo6104 's typescript code.

**Comment by AlbertHart** [comment]
We will need to factor it again after I finish the current changes.

**Comment by sschmidTU** [comment]
The OSMD Transposition plugin is now integrated into OSMD, see https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/733#issuecomment-823530818.
(currently on the develop branch, will be released in the next OSMD version after 0.9.5)
